File: src/state.js
/* eslint-disable no-unused-vars */
/* eslint-disable no-eval */
// TODO Rename this to `$`
const state = {
  classes: [],
};
const _transaction = require("./transaction");
const { graph: _graph } = require("./graph");
const { event } = require("./event");
const _ = require("lodash");
const { v4: uuid } = require("uuid");
const REFERENCE = require("./nuc/REFERENCE");
const serialize = require("./lib/serialize");

global.require = require;

function assign(scope, variable, evaluation, json = true) {
  let value;

  if (json) {
    value = serialize(eval(`(${evaluation})`), "state");
  } else {
    value = evaluation.toString();
  }

  return _transaction.register(`state.${variable}`, value);
}

function call(scope, fn, args = []) {
  const exec = `state.${fn}(${args.join(",")})`;
  return eval(exec);
}

function expression(scope, evaluation) {
  return eval(`(${evaluation.value})`);
}

function del(scope, variable) {
  return eval(`delete state.${variable}`);
}

function load(execs = []) {
  const Node = require("./nuc/NODE");
  const graph = require("./graph");

  execs.forEach((exec) => {
    eval(exec);
  });
}

module.exports.throw = (scope, exception) => eval(`throw ${exception}`);

function clear() {
  for (let property in state) {
    delete state[property];
  }

  state["classes"] = [];
}

module.exports.state = state; // will be private
module.exports.assign = assign;
module.exports.call = call;
module.exports.expression = expression;
module.exports.load = load;
module.exports.clear = clear;
module.exports.delete = del;

File: cpp/src/State.cpp
#include "nucleoid/State.hpp"
#include "nucleoid/JSEngine.hpp"
#include "nucleoid/Transaction.hpp"
#include <sstream>
#include <stdexcept>

namespace nucleoid::state {

nlohmann::json assign(Scope* /*scope*/, const std::string& variable,
                      const nlohmann::json& value) {
    std::string path = "state." + variable;
    return transactionRegister(path, value);
}

nlohmann::json call(Scope* /*scope*/, const std::string& fn,
                    const std::vector<nlohmann::json>& args) {
    std::string code = "state." + fn + "(";
    for (std::size_t i = 0; i < args.size(); ++i) {
        if (i) code += ",";
        code += args[i].dump();
    }
    code += ")";
    return JSEngine::instance().eval(code);
}

nlohmann::json expression(Scope* /*scope*/, const std::string& expr) {
    return JSEngine::instance().eval("(" + expr + ")");
}

bool del(Scope* /*scope*/, const std::string& variable) {
    return JSEngine::instance().eval("delete state." + variable).get<bool>();
}

[[noreturn]] void throwException(Scope* /*scope*/, const std::string& expr) {
    std::string msg;
    try {
        JSEngine::instance().eval(expr); // will throw JSException
        msg = expr;
    } catch (const std::exception& e) {
        msg = e.what();
    }
    throw std::runtime_error(msg);
}

void clear() {
    JSEngine::instance().eval(
        "state = { classes: [] }; $ = { classes: {} };");
}

} // namespace nucleoid::state
